raiden/network/transport/matrix.py

```diff
@@ -573,28 +644,65 @@ def _send_queued_messages(
 
     def _send_with_retry(
         self,
-        receiver_address: Address,
-        async_result: AsyncResult,
-        data: str,
+        queue_identifier: QueueIdentifier,
+        message: Message,
     ):
+        data = json.dumps(message.to_dict())
+        message_id = message.message_identifier
+        receiver_address = queue_identifier.recipient
+        reachable = {UserPresence.ONLINE, UserPresence.UNAVAILABLE}
+
         def retry():
-            if not self._running:
-                return
             timeout_generator = udp_utils.timeout_exponential_backoff(
                 self._config['retries_before_backoff'],
                 self._config['retry_interval'],
                 self._config['retry_interval'] * 10,
             )
-            while async_result.value is None:
-                self._send_immediate(receiver_address, data)
-                gevent.sleep(next(timeout_generator))
+            for delay in timeout_generator:
+                state = self._address_to_presence.get(receiver_address)
+                if state in reachable:
+                    self._send_raw(receiver_address, data)
+                else:
+                    self.log.debug(
+                        'Skipping SEND to unreachable node',
+                        receiver=pex(receiver_address),
+                        state=state,
+                        message=message,
+                        queue=queue_identifier,
+                    )
+                # equivalent of gevent.sleep, but bails out when stopping
+                if self._stop_event.wait(delay):
+                    break
+                # retry while our queue is valid
+                if queue_identifier not in self._queueids_to_queues:
+                    self.log.debug(
+                        'Queue cleaned, stop retrying',
+                        message=message,
+                        queue=queue_identifier,
+                        queueids_to_queues=self._queueids_to_queues,
+                    )
+                    break
+                # retry while the message is in queue
+                # Delivered and Processed messages should eventually remove them
+                if not any(
+                    message_id == event.message_identifier
+                    for event in self._queueids_to_queues[queue_identifier]
```

Comment:
could you move this up so that we don't have a multi line if-statement?

Reply:
It's not a multi-line if, it's a multi-line `any` (I know, I know, just telling my 'justification'). But ok, will do